### PR TITLE
Add PIDFile to docker.service unit file to make systemd clean up pid file more reliably

### PR DESCRIPTION
**Pull request by qiutongs**
**- What I did**
In docker.service unit file, add a PIDFile option with the default pid file path of docker.

**- Why I did**

In the case of a high churn process system there is an edge case where the process is killed and the process id is assigned to a new running process before systemd can restart the process and see that no process is running under the old pid. When the in-use PIDs is close to the system capacity, it's likely that the PID has been reused by some other process. The dockerd does a check to see if the pid file exist and the process id in the file is running[3]. In this case, docker will fail to restart since the pid file contains a valid, running process.

So configuring the PIDFile in docker.service, systemd will help cleanup the pid file more reliably.

**- How I did it**

Updated the docker.service

**- How to verify it**

Try to simulate the high churn behavior.

Before the change:
```
# sudo su
# systemctl stop docker
# echo "1" > /var/run/docker.pid
# systemctl start docker
Job for docker.service failed because the control process exited with error code.
See "systemctl status docker.service" and "journalctl -xe" for details.
# docker ps
error during connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/containers/json: read unix @->/var/run/docker.sock: read: connection reset by peer
# docker ps
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

After the change:
```
# systemctl daemon-reload
# systemctl stop docker
# echo "1" > /var/run/docker.pid
# systemctl start docker
Job for docker.service failed because the control process exited with error code.
See "systemctl status docker.service" and "journalctl -xe" for details.
# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
# systemctl status docker
● docker.service - Docker Application Container Engine
   Loaded: loaded (/etc/systemd/system/docker.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2020-08-27 21:20:30 UTC; 6s ago
     Docs: https://docs.docker.com
  Process: 50532 ExecStopPost=/bin/echo docker daemon exited (code=exited, status=0/SUCCESS)
  Process: 50541 ExecStartPre=/bin/sh -c if [[ -f /var/lib/docker/daemon.json ]]; then cp -f /var/lib/docker/daemon.json /etc/docker/daemon.json; fi (code=exited, status=0/SUCCESS)
 Main PID: 50542 (dockerd)
    Tasks: 8
   Memory: 36.1M
      CPU: 206ms
   CGroup: /system.slice/docker.service
           └─50542 /usr/bin/dockerd --registry-mirror=https://mirror.gcr.io --host=fd:// --containerd=/var/run/containerd/containerd.sock
```

**- Description for the changelog**

Add PIDFile to docker.service unit file to make systemd clean up pid file more reliably

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks! Could you perhaps amend your commit message, and include the information you put in the GitHub PR description in the commit message so that it's preserved in Git history?

Could you also open a pull request with the same change in the docker-ce-packaging repository; https://github.com/docker/docker-ce-packaging/blob/master/systemd/docker.service (as that's the unit file that's currently used as part of the packages)

**Comment by qiutongs** [comment]
> Thanks! Could you perhaps amend your commit message, and include the information you put in the GitHub PR description in the commit message so that it's preserved in Git history?
> 
> Could you also open a pull request with the same change in the docker-ce-packaging repository; https://github.com/docker/docker-ce-packaging/blob/master/systemd/docker.service (as that's the unit file that's currently used as part of the packages)

@thaJeztah  Sure, I can do that. But before that, can you help elaborate on the failed tests?

**Comment by qiutongs** [comment]
What is the command to re-run the tests?

**Comment by thaJeztah** [comment]
Currently has to be done through Jenkins itself. I'll give it another kick, but I we could ignore the failure on s390x if it continues failing.

**Review by thaJeztah — DISMISSED** [review]
LGTM

@tianon PTAL

**Comment by tianon** [comment]
Do you think we're safe to rely on / assume the default setting here, or do we also need to provide `--pidfile` on `ExecStart=`?

(For example, if a user supplies an alternate path in `/etc/docker/daemon.json` from the default value.)

**Comment by qiutongs** [comment]
> Do you think we're safe to rely on / assume the default setting here, or do we also need to provide `--pidfile` on `ExecStart=`?
> 
> (For example, if a user supplies an alternate path in `/etc/docker/daemon.json` from the default value.)

That is a good question.  I thought it is safe because "ExecStart" is also defined within this service unit file. Is user able to supply a config path from external? I might miss something there.

File: contrib/init/systemd/docker.service [hunk 1]
```diff
@@ -10,7 +10,8 @@ Type=notify
 # the default is not to use systemd for cgroups because the delegate issues still
 # exists and systemd currently does not support the cgroup feature set required
 # for containers run by docker
-ExecStart=/usr/bin/dockerd -H fd://
+# explicitly set the PID file path to be aligned with PIDFile option in this file 
+ExecStart=/usr/bin/dockerd -H fd:// --pidfile /var/run/docker.pid
```
Comment (thaJeztah) [inline contrib/init/systemd/docker.service:14]: We should probably also add the containerd socket here; like in https://github.com/docker/docker-ce-packaging/blob/master/systemd/docker.service#L14

```suggestion
ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock --pidfile /var/run/docker.pid
```

(see https://github.com/docker/docker-ce-packaging/pull/297 for details on that)

**Review by thaJeztah — COMMENTED** [review]
> That is a good callout. I thought it is safe because I didn't consider the scenario you described where user can configure dockerd externally. So I agree with you that it is better to enforce the --pidfile

So, wondering, there's also the `docker.socket` unit-file (for socket activation); which one of the unit files should check for the pidfile? https://github.com/moby/moby/blob/master/contrib/init/systemd/docker.socket

And the unit-file for RPM's (which in this repository was separate, without socket activation); https://github.com/moby/moby/blob/master/contrib/init/systemd/docker.service.rpm

**Comment by thaJeztah** [comment]
Looks like this is a duplicate of https://github.com/moby/moby/pull/43924, on which also no consensus was reached. I originally gave my "LGTM", but I agree that I don't see an immediate value on doing this, so unless there's a strong use-case for this, I don't think we should make this change.

I'll close this PR for now (so that's a no _for now_, but we can still change our mind in future, pending more information); feel free to continue the conversation though after I closed.

**Pull request closed without merging**